File: BOJ/[1932] 정수 삼각형/kw601/solution.py
import sys

inputs = sys.stdin.read().splitlines()

n = int(inputs[0])
triangle = [list(map(int, l.split())) for l in inputs[1:]]

# dp 행렬 초기화
dp = [row[:] for row in triangle]

for i in range(len(triangle) - 2, -1, -1):
    for j in range(len(triangle[i])):
        try:
            dp[i][j] = triangle[i][j] + max(dp[i + 1][j], dp[i + 1][j + 1])
        except:
            continue
print(dp[0][0])

File: BOJ/[1932] 정수 삼각형/kw601/solution.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        List<int[]> triangleList = new ArrayList<>();

        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            String[] line = br.readLine().split(" ");
            int[] row = new int[line.length];
            for (int j = 0; j < line.length; j++) {
                row[j] = Integer.parseInt(line[j]);
            }
            triangleList.add(row);
        }

        int[][] triangle = triangleList.toArray(new int[0][]);
        
        // 복사를 이렇게 해야 한대...
        int[][] dp = new int[n][];
        for (int i = 0; i < n; i++) {
            dp[i] = triangle[i].clone();
        } 

        for (int i = n - 2; i >= 0; i--) {
            for (int j = 0; j < triangle[i].length; j++) {
                dp[i][j] = triangle[i][j] + Math.max(dp[i + 1][j], dp[i + 1][j + 1]);
            }
        }

        System.out.println(dp[0][0]);
    }
}